### PR TITLE
Security issues

### PR DESCRIPTION
**Pull request by karen-avetisyan-mc**
#### Description
Improving the certificate loading and removing the technical debt.
pyOpenSSL library is completely removed.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -16,11 +16,10 @@
...
           'Operating System :: OS Independent',
-          'Programming Language :: Python :: 3.6',
-          'Programming Language :: Python :: 3.7',
           'Programming Language :: Python :: 3.8',
+          'Programming Language :: Python :: 3.9',
           'Topic :: Software Development :: Libraries :: Python Modules'
       ],
       tests_require=['coverage'],
-      install_requires=['pycryptodome>=3.8.1', 'pyOpenSSL>=22.1.0,<=23.2.0', 'setuptools>=39.0.1' ]
+      install_requires=['pycryptodome>=3.8.1', 'setuptools>=69.1.0',"cryptography>=42.0.0" ]
```
Comment (Akshaykdubey) [inline setup.py:24]: `install_requires=['pycryptodome>=3.8.1', 'setuptools>=69.1.0', "cryptography>=42.0.0" ]`

File: client_encryption/field_level_encryption_config.py [hunk 2]
```diff
@@ -24,19 +25,20 @@ def __init__(self, conf):
             self._paths[path] = EncryptionPathConfig(opt)
 
         if "encryptionCertificate" in json_config:
-            x509_cert = load_encryption_certificate(json_config["encryptionCertificate"])
-            self._encryption_certificate = dump_certificate(FILETYPE_ASN1, x509_cert)
+            x509_cert, cert_type = load_encryption_certificate(json_config["encryptionCertificate"])
+            self._encryption_certificate = x509_cert
+            #Fixed encoding is required, regardless of initial cerrtificate encoding to ensure correct calcualtion of fingerprint value
```
Comment (Akshaykdubey) [inline client_encryption/field_level_encryption_config.py:30]: `# Fixed encoding is required, regardless of initial certificate encoding to ensure correct calculation of fingerprint value`

File: client_encryption/field_level_encryption_config.py [hunk 3]
```diff
@@ -1,8 +1,9 @@
 import json
-from OpenSSL.crypto import dump_certificate, FILETYPE_ASN1, dump_publickey
 from Crypto.Hash import SHA256
-from client_encryption.encoding_utils import Encoding
+from client_encryption import encoding_utils
 from client_encryption.encryption_utils import load_encryption_certificate, load_decryption_key, validate_hash_algorithm
+from cryptography.hazmat.primitives.serialization import PublicFormat, Encoding
+
```
Comment (Akshaykdubey) [inline client_encryption/field_level_encryption_config.py:6]: This file needs formatting

File: client_encryption/encryption_utils.py [hunk 4]
```diff
@@ -1,26 +1,47 @@
...
-
-        return x509
     except IOError:
-        raise CertificateError("Unable to load certificate.")
-    except (ValueError, Error):
-        raise CertificateError("Wrong encryption certificate format.")
+        raise CertificateError ("Unable to load certificate.")
+    
+    try:         
+        type = __get_crypto_file_type(certificate)
```
Comment (Akshaykdubey) [inline client_encryption/encryption_utils.py:29]: Don't use `type`, it shadows a builtin function.

File: client_encryption/encryption_utils.py [hunk 5]
```diff
@@ -1,26 +1,47 @@
...
+            cert = x509.load_pem_x509_certificate(certificate)
+            return cert, Encoding.PEM
+        if type == FileType.FILETYPE_ASN1:
+            cert = x509.load_der_x509_certificate(certificate)
+            return cert, Encoding.DER
+        if type == FileType.FILETYPE_INVALID:
+            raise CertificateError("Wrong certificate format.")
+    except ValueError:
+            raise CertificateError ("Invalid  certificate format.")
+
```
Comment (Akshaykdubey) [inline client_encryption/encryption_utils.py:41]: This file needs formatting

File: client_encryption/encryption_utils.py [hunk 6]
```diff
@@ -1,26 +1,47 @@
 from Crypto.PublicKey import RSA
 from Crypto.Hash import SHA1, SHA224, SHA256, SHA384, SHA512
-from OpenSSL.crypto import load_certificate, FILETYPE_PEM, FILETYPE_ASN1, Error
+#from OpenSSL.crypto import load_certificate, FILETYPE_PEM, FILETYPE_ASN1, Error
```
Comment (Akshaykdubey) [inline client_encryption/encryption_utils.py:3]: Remove this line

File: client_encryption/encryption_utils.py [hunk 7]
```diff
@@ -1,26 +1,47 @@
...
 from client_encryption.encryption_exception import CertificateError, PrivateKeyError, HashAlgorithmError
+from cryptography import x509
 from cryptography.hazmat.primitives.serialization import pkcs12 
 from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.serialization import Encoding
+from enum import IntEnum
 
 _SUPPORTED_HASH = {"SHA1": SHA1, "SHA224": SHA224, "SHA256": SHA256, "SHA384": SHA384, "SHA512": SHA512}
```
Comment (Akshaykdubey) [inline client_encryption/encryption_utils.py:12]: This file needs formatting

File: client_encryption/encryption_utils.py (hunk 4 above)
Comment (karen-avetisyan-mc) [reply client_encryption/encryption_utils.py]: done

File: client_encryption/jwe_encryption_config.py [hunk 8]
```diff
@@ -26,15 +26,16 @@ def __init__(self, conf):
             self._paths[path] = EncryptionPathConfig(opt)
 
         if "encryptionCertificate" in json_config:
-            x509_cert = load_encryption_certificate(json_config["encryptionCertificate"])
-            self._encryption_certificate = dump_certificate(FILETYPE_ASN1, x509_cert)
+            x509_cert, cert_type = load_encryption_certificate(json_config["encryptionCertificate"])
+            self._encryption_certificate = x509_cert
+            #Fixed encoding is required, regardless of initial cerrtificate encoding to ensure correct calcualtion of fingerprint value
```
[the remark this answers is not in the document]
Comment (karen-avetisyan-mc) [reply client_encryption/jwe_encryption_config.py]: fixed

**Pull request merged**